File: node/PostOffice/src/features/support/CreateLabelInvalidDimensionsSteps.ts
import {Given, Then, When} from "@cucumber/cucumber";
import {DeliveryType} from "../../models/DeliveryType";
import {Package} from "../../models/Package";
import {LabelService} from "../../services/LabelService";
import assert from "assert";


let pkg: Package;
let error: Error;

Given('a package with invalid dimensions {int} height, {int} width, {int} length, and weight {float} kg, and delivery type {string}', function (height, width, length, weight, deliveryType) {
  const deliveryEnum = DeliveryType[deliveryType as keyof typeof DeliveryType];
  if (!deliveryEnum) {
    throw new Error(`Invalid delivery type: ${deliveryType}`);
  }
  pkg = new Package("Id", height, width, length, weight, deliveryEnum);
});


When('the label is created for the invalid dimensions package', function () {
  try {
    LabelService.createLabel(pkg);
  } catch (e) {
    error = e as Error;
  }
});

Then('an exception should be thrown with {string}', function (expectedMessage) {
  assert.equal(error.message, expectedMessage);
});

File: java/cucumber-archetype/src/test/java/postoffice/stepdefs/CreateLabelInvalidDimensionsSteps.java
package postoffice.stepdefs;

import io.cucumber.java.en.Given;
import io.cucumber.java.en.When;
import io.cucumber.java.en.Then;
import static org.junit.Assert.*;

import postoffice.models.Package;
import postoffice.models.DeliveryType;
import postoffice.services.LabelService;

public class CreateLabelInvalidDimensionsSteps {

    private Package pkg;
    private LabelService labelService = new LabelService();
    private String errorMessage;

    @Given("a package with invalid dimensions {int} height, {int} width, {int} length, and weight {double} kg, and delivery type {string}")
    public void createLabelInvalidDimensionsPackage(int height, int width, int length, double weight,
            String deliveryType) {
        pkg = new Package(); // Initialize a new Package object

        pkg.setHeight(height);
        pkg.setWidth(width);
        pkg.setLength(length);
        pkg.setWeight(weight);
        pkg.setDeliveryType(DeliveryType.valueOf(deliveryType.toUpperCase())); // Convert string to DeliveryType enum
    }

    @When("the label is created for the invalid dimensions package")
    public void invalidDimensions_whenTheLabelIsCreated() {
        try {
            labelService.createLabel(pkg);
        } catch (IllegalArgumentException e) {
            errorMessage = e.getMessage(); // Capture the exception message
        }
    }

    @Then("an exception should be thrown with {string}")
    public void thenAnExceptionShouldBeThrownDueToInvalidDimensions(String expectedMessage) {
        assertEquals(expectedMessage, errorMessage);
    }
}